Tidy up code

alexdewar opened the pull request:
This is a collection of misc tidy-ups I've made to the code. I'm not sure whether we necessarily want to merge this before the course is delivered on Fri; we can just leave it as is for now. (I'm also away tomorrow so won't be able to work on the course then.)

Once this is merged, I'll sync the content in the lessons with this repo, possibly as part of this issue: https://github.com/ImperialCollegeLondon/unit_testing_Cpp/issues/30

There are some other small improvements we could make to the code, but I think these are the main issues.

Commit summary:

- Move a couple of constants defined in `employee.h` to `employee.cpp` (you shouldn't have variable definitions in headers)
- Add underscore suffix to member variables and use camel case consistently (generally people either add an underscore to member variable names or prefix them with `m_` to distinguish them from local variables)
- Remove unused `Employee` destructor
- Merge duplicate `employee.h` and `employee.cpp` files
- Other small clean-ups

dalonsoa reviewed (approved):
These changes look good. The fact that you have been able pinpoint these changes suggest that there is a sort of standard of good practices in C++, right? Is this clang-tidy related or coming from somewhere else?

alexdewar commented:
@lkr-ind I renamed the member variables manually and opted to make them all camel case for consistency, as we previously had a mixture of camel case and snake case. As camel case is used everywhere else it seemed like the easiest option. I don't prefer any particular style, but we should be consistent.

I also changed the types of the `age_` and `numberOfYears_` members to `unsigned int` manually (though I forgot to mention this in the PR description -- sorry!). Looking through the code it seems that we are only ever assigning integers to these members anyway, so it seems clearer to just use an integer type.

jamesturner246 commented on src/Chapter3/employee_table.cpp:
```diff
@@ -5,30 +5,30 @@
-        if (it->getName() == employeeName) {
-            employees.erase(it);
+void EmployeeTable::removeEmployee(const std::string &name) {
+    for (auto it = employees_.begin(); it != employees_.end(); ++it) {
+        if (it->getName() == name) {
+            employees_.erase(it);
             break;
         }
     }
```
```suggestion
std::erase_if(employees_, [](const Employee &e){e.getName() == name;});
```
C++20 ftw!

alexdewar replied:
I'm afraid this repo is stuck on C++14, at least for now. I'd *mayyyybe* consider bumping it to C++17 if we had a use for it, but I'm a bit concerned that C++20 might be a bit too bleeding edge for some of the compilers students will have installed :disappointed:.

It's surprising that `std::erase_if` took so long to make it into the standard though!

jamesturner246 replied:
Ah, I see. We could do the following, since C++11:

```cpp
emplloyees_.erase(
    std::remove_if(employees_.begin(), employees_.end(),
    [](const Employee &e){e.getName() == name;}), employees_.end();
```
Just doesn't look as nice :(

alexdewar replied:
Yeah. C++ often ends up looking rather ugly :disappointed:

lkr-ind commented:
@alexdewar Thanks for the clarifications. 

I have a small difference of opinion (just on the variable type) . Let me know if you think otherwise.

I think it is a good idea to assign the variable types based on what could be used for those values rather than what has been used in the small sample code. For example, age could be any floating point number. So, I would be inclined towards `float` or `double` ( An additional check might be necessary for negative values in this case).

jamesturner246 commented on src/Chapter4/2_test_using_for_loop.cpp:
```diff
@@ -9,7 +9,7 @@ class EmployeeTestFixture : public ::testing::Test {
 TEST_F(EmployeeTestFixture, NetBonusIsCorrectForDifferentYears) {
-    auto input = std::vector<int>{5, 15};
+    auto input = std::vector<unsigned int>{5, 15};
     auto expected_output = std::vector<int>{2000, 3000};
```
Shouldn't outputs be `float`?

alexdewar replied:
Yes! Good spot.

And we should be using `EXPECT_DOUBLE_EQ` below instead.

jamesturner246 commented on src/Chapter4/3_parameterised_not_using_fixture.cpp:
```diff
@@ -4,11 +4,11 @@
 struct TestValues {
-    int input;
+    unsigned int input;
     int output;
```
Same.

jamesturner246 commented on src/Chapter4/3_parameterised_not_using_fixture.cpp:
```diff
@@ -4,11 +4,11 @@
     // constructor of values struct
-    TestValues(int in, int out) : input(in), output(out) {
+    TestValues(unsigned int in, int out) : input(in), output(out) {
     }
```
Same.

alexdewar commented:
@dalonsoa There are *sort* of best practices for C++, though different people use different styles for certain things. For example, in some code bases, member variable names are prefixed with `m` and in others an underscore is added (e.g. in [Google's style guide](https://google.github.io/styleguide/cppguide.html#Variable_Names)). The closest thing I know of to a standard set of best practices is the [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines), written by the original author of C++, but there are a lot of things it doesn't cover (confession: I haven't actually read it).

*Some* of the changes I made are things that `clang-tidy` can warn you about. For example, the `getName()` function's return type was `std::string`, which meant that a new string object is created every time you call it. You're not likely to actually want to modify this though, so it makes more sense to return a `const std::string &`. If you do that, then you can *still* create a new string object *if you want*, in the same way, by writing:

```cpp
std::string name = employee.getName();
```

This will copy from the `const` reference and make a new object for you. Note that there's no advantage to *not* returning a `const` reference here! It just means you don't *have* to copy the object, unless you want to -- a free performance win. So that's why tools like `clang-tidy` will pick you up on it and why it's best practice to return objects from getters as `const` ref. (The rule doesn't apply for basic types like `int` because it's small enough to fit in a single CPU register, so there's no advantage to copying a reference rather than itself.) Of course, in this case the performance advantage would be absolutely minuscule, so doing it this way is more of a code hygiene thing, but if you were returning a `std::vector<double>` with 1,000,000 elements or something the performance impact could be pretty considerable.

jamesturner246 reviewed (changes requested):
Cool. Some minor things.

Just curious: is the camelCase for class state, underscore for other, intentional?

Ed: ah I see your answer above. NVM

alexdewar commented:
@lkr-ind Yes, you might want to represent someone's age in terms of fractions of a year. (Though I'm not sure why age was a `float` when everything else was a `double`?)

I think it's more intuitive for age to be an integer though, because people talk about ages in terms of whole years rather than fractions. In any case, for this example code, we only ever use integers for ages, so future-proofing the code for a scenario where someone *might* want to represent a fractional age just confuses things, I think.

jamesturner246 reviewed (approved):
No problem. I'd say go ahead then.

alexdewar commented:
I actually found a bunch of other places that we were doing `EXPECT_EQ` where we should have been doing `EXPECT_DOUBLE_EQ` instead. All the more reason to avoid using floating-point numbers where you can :wink:

As I said, I'm away tomorrow and I'm conscious that @jamesturner246 and @lkr-ind are delivering on Fri, so I'm happy to hold off on merging this until that's done. In the meantime, I was planning on adding setup instructions to the course material.
